algorithms/src/snark/marlin/data_structures/proof.rs

```diff
@@ -131,99 +134,136 @@ impl<F: PrimeField> Evaluations<F> {
         mut writer: W,
         compress: Compress,
     ) -> Result<(), snarkvm_utilities::SerializationError> {
-        for z_b_eval in &self.z_b_evals {
-            CanonicalSerialize::serialize_with_mode(z_b_eval, &mut writer, compress)?;
+        for z_b_eval_circuit in &self.z_b_evals {
+            serialize_vec_without_len(z_b_eval_circuit, &mut writer, compress)?;
         }
         CanonicalSerialize::serialize_with_mode(&self.g_1_eval, &mut writer, compress)?;
-        CanonicalSerialize::serialize_with_mode(&self.g_a_eval, &mut writer, compress)?;
-        CanonicalSerialize::serialize_with_mode(&self.g_b_eval, &mut writer, compress)?;
-        CanonicalSerialize::serialize_with_mode(&self.g_c_eval, &mut writer, compress)?;
+        serialize_vec_without_len(&self.g_a_evals, &mut writer, compress)?;
+        serialize_vec_without_len(&self.g_b_evals, &mut writer, compress)?;
+        serialize_vec_without_len(&self.g_c_evals, &mut writer, compress)?;
         Ok(())
     }
 
     fn serialized_size(&self, compress: Compress) -> usize {
         let mut size = 0;
-        size += self.z_b_evals.iter().map(|s| s.serialized_size(compress)).sum::<usize>();
+        for z_b_eval_circuit in &self.z_b_evals {
+            size += serialized_vec_size_without_len(z_b_eval_circuit, compress);
+        }
         size += CanonicalSerialize::serialized_size(&self.g_1_eval, compress);
-        size += CanonicalSerialize::serialized_size(&self.g_a_eval, compress);
-        size += CanonicalSerialize::serialized_size(&self.g_b_eval, compress);
-        size += CanonicalSerialize::serialized_size(&self.g_c_eval, compress);
+        size += serialized_vec_size_without_len(&self.g_a_evals, compress);
+        size += serialized_vec_size_without_len(&self.g_b_evals, compress);
+        size += serialized_vec_size_without_len(&self.g_c_evals, compress);
         size
     }
 
     fn deserialize_with_mode<R: snarkvm_utilities::Read>(
-        batch_size: usize,
+        batch_sizes: &[usize],
         mut reader: R,
         compress: Compress,
         validate: Validate,
     ) -> Result<Self, snarkvm_utilities::SerializationError> {
-        let mut z_b_evals = Vec::with_capacity(batch_size);
-        for _ in 0..batch_size {
-            z_b_evals.push(CanonicalDeserialize::deserialize_with_mode(&mut reader, compress, validate)?);
-        }
         Ok(Evaluations {
-            z_b_evals,
+            z_b_evals: batch_sizes
+                .iter()
+                .map(|batch_size| deserialize_vec_without_len(&mut reader, compress, validate, *batch_size))
+                .collect::<Result<Vec<Vec<_>>, _>>()?,
```

Comment:
```suggestion
                .collect::<Result<_, _>>()?,
```
type inference should be able to handle this

Reply:
Good catch cef8f291b